### ISSUE TITLE
[BUG]  when i run stm32f103** connect with smartcard,why it always printf" Error:..\FreeRTOS\portable\RVDS\ARM_CM3\port.c,378"

### ISSUE THREAD

**Issue by Guoina** [FreeRTOS/FreeRTOS-Kernel#423]
**Describe the bug**
When i run stm32f103** connect with smartcard, why it always printf" Error:..\FreeRTOS\portable\RVDS\ARM_CM3\port.c,378".the freertos version is FreeRTOS V9.0.0

**Target**
- Development board: [stm32f103ret6]
- Instruction Set Architecture: [Cortex-M3]
- IDE and version: [uvision v5.23.0.0 ]
- Toolchain and version: [MDK-ARM PLUS version 5.23]

windows

**To Reproduce**
- Use project ... and configure with ...
- Run on ... and could observe ...

**Expected behavior**
it never happens

**Screenshots**

![image](https://user-images.githubusercontent.com/68848066/146133098-3b1b2bb6-f8a5-4bec-b21a-05b8ec0a6537.png)

**Comment by paulbartell** [comment]
@Guoina, A copy of the source code for that assertion is located over here: https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/V9.0.0/FreeRTOS/Source/portable/RVDS/ARM_CM3/port.c#L370 

The associated note is copied below:
```
	/* This is not the interrupt safe version of the enter critical function so
	assert() if it is being called from an interrupt context.  Only API
	functions that end in "FromISR" can be used in an interrupt.  Only assert if
	the critical nesting count is 1 to protect against recursive calls if the
	assert function also uses a critical section. */
```

Generally, we suggest that runtime assertions in debug builds cause the application to stop altogether rather than continue running.

In short, FreeRTOS is working as-designed and your driver is likely calling the [taskENTER_CRITICAL](https://www.freertos.org/taskENTER_CRITICAL_taskEXIT_CRITICAL.html) macro from an ISR where it should be calling the [taskENTER_CRITICAL_FROM_ISR](https://www.freertos.org/taskENTER_CRITICAL_FROM_ISR_taskEXIT_CRITICAL_FROM_ISR.html) macro instead.

Feel free to reopen this ticket if you have any further questions.

**Comment by RichardBarry** [comment]
As per the instructions shown when opening a github issue (screenshot below), please use the support forum for support requests (https://forums.freertos.org), and only open bugs when an issue has been determined to be a bug.  Thanks.

![image](https://user-images.githubusercontent.com/3073890/146294731-876ac856-4362-4082-9f30-b1a5ec85e5d3.png)

**Comment by Guoina** [comment]
Thans for your reply,and sorry for my "mistake bug ",I will check my code carefully. 










At 2021-12-16 04:02:26, "Paul Bartell" ***@***.***> wrote:

@Guoina, A copy of the source code for that assertion is located over here: https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/V9.0.0/FreeRTOS/Source/portable/RVDS/ARM_CM3/port.c#L370

The associated note is copied below:

	/* This is not the interrupt safe version of the enter critical function so
	assert() if it is being called from an interrupt context.  Only API
	functions that end in "FromISR" can be used in an interrupt.  Only assert if
	the critical nesting count is 1 to protect against recursive calls if the
	assert function also uses a critical section. */


Generally, we suggest that runtime assertions in debug builds cause the application to stop altogether rather than continue running.

In short, FreeRTOS is working as-designed and your driver is likely calling the taskENTER_CRITICAL macro from an ISR where it should be calling the taskENTER_CRITICAL_FROM_ISR macro instead.

Feel free to reopen this ticket if you have any further questions.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
